add interfaces flag with unit test

kliukovkin opened the pull request:
This is a proposal for `-interfaces` flag using source mode. Using this flag it is possible to list only required interfaces to be mocked in source mode. Currently, source mode mocks all interfaces which -source file contains.

Thanks @KastenMike for raising this!

Fixes https://github.com/golang/mock/issues/660

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber-go/mock?pullRequest=200) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber-go/mock?pullRequest=200) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber-go/mock?pullRequest=200) it.</sub>

KastenMike commented on mockgen/mockgen.go:
```diff
@@ -115,6 +116,14 @@ func main() {
 		return
 	}
 
+	if len(*interfaces) > 0 {
```
maybe this should go in `sourceMode` like `p.excludeNamesSet = parseExcludeInterfaces(*excludeInterfaces)`?

kliukovkin commented:
Hey @JacobOaks , can you please take a look?

JacobOaks commented:
Hey @kliukovkin! Thanks for the PR!

I am in favor of having this feature, just two questions/comments:
* Instead of using an `-interfaces` flag, what do you think about having source mode read positional arguments (i.e., `mockgen -source <path/to/file.go> InterfaceOne,InterfaceTwo`) to align it with how reflect mode works? (If there are no positional arguments specified, we would parse all interfaces to keep backwards compatibility)
* Instead of parsing and then dropping interfaces that aren't specified, can we simply not parse ones that aren't requested? This is similar to [how the exclusion flag already works](https://github.com/uber-go/mock/blob/main/mockgen/parse.go#L231-L233) and would avoid some wasted computation.

kliukovkin commented:
> Hey @kliukovkin! Thanks for the PR!
> 
> 
> 
> I am in favor of having this feature, just two questions/comments:
> 
> * Instead of using an `-interfaces` flag, what do you think about having source mode read positional arguments (i.e., `mockgen -source <path/to/file.go> InterfaceOne,InterfaceTwo`) to align it with how reflect mode works? (If there are no positional arguments specified, we would parse all interfaces to keep backwards compatibility)
> 
> * Instead of parsing and then dropping interfaces that aren't specified, can we simply not parse ones that aren't requested? This is similar to [how the exclusion flag already works](https://github.com/uber-go/mock/blob/main/mockgen/parse.go#L231-L233) and would avoid some wasted computation.

Hey @JacobOaks ! Done, can you please review and let me know if that looks good?

kliukovkin commented:
@JacobOaks hey Jacob! Gentle ping here.

r-hang commented on mockgen/parse_test.go:
```diff
@@ -143,3 +146,127 @@ func TestParseArrayWithConstLength(t *testing.T) {
+			wantErr: true,
+		},
+		{
+			name: "missing interface (Baz not found)",
```
Could we also add a test case for accidentally specified duplicate requested interfaces?

r-hang commented on mockgen/parse_test.go:
```diff
@@ -143,3 +146,127 @@ func TestParseArrayWithConstLength(t *testing.T) {
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := filterInterfaces(tt.args.all, tt.args.requested)
+			if (err != nil) != tt.wantErr {
```
instead of wantErr being a bool, could we make this a string and compare the error contents?

r-hang reviewed (approved):
Sorry for the delay. Looks good to me with some minor comments. Let's merge this when they are addressed.

abhinav commented on mockgen/parse.go:
```diff
@@ -61,6 +63,20 @@ func sourceMode(source string) (*model.Package, error) {
 	}
 
+	// positional interface names -> include set
+	ifaces := flag.Args()
+	if len(ifaces) > 0 {
+		p.includeNamesSet = make(map[string]struct{}, len(ifaces))
+		for _, arg := range ifaces {
+			for _, name := range strings.Split(arg, ",") {
+				name = strings.TrimSpace(name)
+				if name != "" {
+					p.includeNamesSet[name] = struct{}{}
+				}
+			}
+		}
+	}
```
Other modes accept list of interfaces as a single argument, so I would recommend:

```suggestion
	if flag.NArg() > 1 {
		return nil, errors.New("-source mode accepts at most one argument")
	}
	if flag.NArg() == 1 {
		ifaces := strings.Split(flag.Arg(0), ",")
		p.includeNamesSet = make(map[string]struct{}, len(ifaces))
		for _, name := range ifaces {
			p.includeNamesSet[name] = struct{}{}
		}
	}
```

abhinav commented on mockgen/parse.go:
```diff
@@ -168,6 +185,7 @@ type fileParser struct {
 	auxInterfaces      *interfaceCache
 	srcDir             string
 	excludeNamesSet    map[string]struct{}
+	includeNamesSet    map[string]struct{}
```
worth adding a comment here:

```suggestion
	includeNamesSet    map[string]struct{} // empty to include all
```

abhinav commented on mockgen/parse.go:
```diff
@@ -228,17 +246,36 @@ func (p *fileParser) parseFile(importPath string, file *ast.File) (*model.Packag
 		}
-		i, err := p.parseInterface(ni.name.String(), importPath, ni)
+
+		if len(p.includeNamesSet) > 0 {
+			if _, ok := p.includeNamesSet[name]; !ok {
+				continue
+			}
+		}
```
```suggestion
		// All interfaces are included if no filter was specified.
		if len(p.includeNamesSet) > 0 {
			if _, ok := p.includeNamesSet[name]; !ok {
				continue
			}
		}
```

abhinav commented on mockgen/parse.go:
```diff
@@ -228,17 +246,36 @@ func (p *fileParser) parseFile(importPath string, file *ast.File) (*model.Packag
+		delete(p.includeNamesSet, name)
+	}
+	if len(p.includeNamesSet) > 0 {
+		missing := make([]string, 0, len(p.includeNamesSet))
+		for n := range p.includeNamesSet {
+			missing = append(missing, n)
+		}
+		sort.Strings(missing)
```
We can shorten this:
```suggestion
		missing := slices.Sorted(maps.Keys))
```

(assuming import slices and maps.)

abhinav reviewed (approved):
@sywhang This PR implements the changes you suggested in #281 (failing on missing expected interfaces).

@kliukovkin, thanks for updating. left some minor suggestions for improvements.
I see that the tests are failing, though.

kliukovkin commented:
Updated according to feedback.
- accepted all your suggestions
- removed the “missing interfaces” error check, because it was breaking integration tests - in practice, missing names should just be ignored

Thanks for the review @abhinav ! 🚀

sywhang reviewed (approved):
Thank you!
